core/com.b2international.snowowl.core/src/com/b2international/snowowl/core/request/resource/BaseMetadataResourceConverter.java

```diff
@@ -220,6 +231,115 @@ protected final void expandDependencyUpgrades(List<R> results) {
+			return;
+		}
+		
+		// extensionOf branches are the parent branches of the CodeSystem, so simple branch state calculation is enough
+		for (R result : results) {
+			if (!(result instanceof CodeSystem)) {
+				continue;
+			}
+			
+			CodeSystem codeSystem = (CodeSystem) result;
+			Branch branch = RepositoryRequests.branching().prepareGet(codeSystem.getBranchPath()).build(codeSystem.getToolingId()).execute(context());
+			BranchState branchState = branch.state();
+			codeSystem.setExtensionOfBranchInfo(new BranchInfo(branch.path(), branchState, branch.baseTimestamp(), branch.headTimestamp()));				
+		}
+	}
+
+	private void expandUpgradeOfInfo(List<R> results) {
+		if (!expand().containsKey(CodeSystem.Expand.UPGRADE_INFO)) {
+			return;
+		}
+		
+		final List<ResourceURI> upgradeOfURIs = results.stream()
+				.filter(resource -> resource instanceof CodeSystem)
+				.map(r -> (CodeSystem) r)
+				.filter(codeSystem -> codeSystem.getUpgradeOf() != null)
+				.map(codeSystem -> codeSystem.getUpgradeOf().withoutPath())
+				.collect(Collectors.toList());
+		
+		// nothing to expand, quit early
+		if (upgradeOfURIs.isEmpty()) {
+			return;
+		}
+		
+		final List<String> upgradeOfBranches = context().service(ResourceURIPathResolver.class).resolve(context(), upgradeOfURIs);
+		
+		final Map<ResourceURI, String> branchesByUpgradeOf = Maps.newHashMap();
+		Iterator<ResourceURI> uriIterator = upgradeOfURIs.iterator();
+		Iterator<String> branchIterator = upgradeOfBranches.iterator();
+		while (uriIterator.hasNext() && branchIterator.hasNext()) {
+			ResourceURI uri = uriIterator.next();
+			String branch = branchIterator.next();
+			branchesByUpgradeOf.put(uri, branch);
+		}
+
+		for (R result : results) {
+			
+			if (!(result instanceof CodeSystem)) {
+				continue;
+			}
+			
+			CodeSystem cs = (CodeSystem) result;
+			if (cs.getUpgradeOf() == null) {
+				continue;
+			}
+			String upgradeOfCodeSystemBranchPath = branchesByUpgradeOf.get(cs.getUpgradeOf().withoutPath());
+			
+			if (!Strings.isNullOrEmpty(upgradeOfCodeSystemBranchPath)) {
+				RepositoryContext ctx = context().service(RepositoryManager.class).getContext(cs.getToolingId());
+				BaseRevisionBranching branching = ctx.service(BaseRevisionBranching.class);
+				
+				RevisionBranch branch = branching.getBranch(cs.getBranchPath());
+				BranchState branchState = branching.getBranchState(cs.getBranchPath(), upgradeOfCodeSystemBranchPath);
+				BranchInfo mainInfo = new BranchInfo(branch.getPath(), branchState, branch.getBaseTimestamp(), branch.getHeadTimestamp());
+				
+				List<ResourceURI> availableVersions = Lists.newArrayList();
+				List<BranchInfo> versionBranchInfo = Lists.newArrayList();
+
+				if (!cs.getUpgradeOf().isHead()) {
+					long startTimestamp;
+					final String upgradeOfVersionBranch = context().service(ResourceURIPathResolver.class).resolve(context(), List.of(cs.getUpgradeOf())).stream().findFirst().orElse("");
```

Comment:
There is a simplified version of resolve() in ResourceURIPathResolver which does not need a List of URIs as an input: `com.b2international.snowowl.core.uri.ResourceURIPathResolver.resolve(ServiceProvider, ResourceURI, Resource)`